### ISSUE TITLE
Is there a way to express the concept of "buy/sell the next bar"?

### ISSUE THREAD

**Issue by sixro** [ta4j/ta4j#521]
**I have checked existing issues and wiki**
- [x] I could not find similar [issues](https://github.com/ta4j/ta4j/issues?utf8=%E2%9C%93&q=)
- [x] I could not find a solution in the [wiki](https://ta4j.github.io/ta4j-wiki/) or [faq section](https://ta4j.github.io/ta4j-wiki/FAQ.html)

If possible provide a **[minimal working example](https://stackoverflow.com/help/mcve)**!

Hi everyone, trying to make a backtest of a strategy I already have (I'm using ProRealtime), I found an issue that I cannot fix apparently.
Basically, my strategy should buy at market on the next bar.
It seems to me that at the moment the `BarSeriesManager` operate on the same bar where the entry rule is applied. I tried to play with the `WaitForRule`, but without any success.
Is it something known?
Do you suggest to create a specific `BarSeriesManager` to handle this?
Regards
  R

**Comment by nimo23** [comment]
The WaitForRule has the property `numberOfBars`: The number of bars to wait for

Set this bar to "1" ..

**Comment by sixro** [comment]
Hi @nimo23 and thanks.
I'll try again with the `WaitForRule`, but I don't think it works combining it with another `Rule` with `AndRule` for the entry: because when the wait-for rule will be true the other one could be false...
Anyway another problem remains: how to buy at market.
For me this means, that on daily bars the open price of the next bar should be used.
Reading the code of `BarSeriesManager`, I think only the close price on the "selected" bar is used.

**Comment by nimo23** [comment]
>because when the wait-for rule will be true the other one could be false...

that's right. I think the ChainRule would be good for that: https://github.com/ta4j/ta4j/pull/472.

**Comment by team172011** [comment]
merged

**Comment by sixro** [comment]
Not sure I can comment again on a closed issue, but after some weeks I can give you a feedback: I decided to create a new `BarSeriesManager` and an extension of `Strategy` where the strategy is responsible to decide the entry type to use (at market? with a limit/stop order?).
The `BarSeriesManager` has to process this entry type correctly waiting for the next bar.
The main reason behind this change is that waiting for the next bar was wrong because the system was purchasing anyway at the close of the next bar and it was not correct.
Unfortunately the code is totally against the design behind the library, full of tricks, etc... so it is not candidate to a pull request...

**Comment by nimo23** [comment]
>The main reason behind this change is that waiting for the next bar using the WaitForRule was wrong because the system was purchasing anyway at the close of the next bar and it was not correct.

If so then please create a new issue for this to find out if there is a bug or not. 

Anyway, WaitForRule and ChainRule are different things..you should try it with ChainRule instead.
